src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs

```diff
@@ -3,16 +3,214 @@
+                typeof(T) == typeof(nuint);
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static bool CompareExchange(ref T location, T value, T comparand) =>
+                typeof(T) == typeof(int) ? Interlocked.CompareExchange(ref Unsafe.As<T, int>(ref location), Unsafe.As<T, int>(ref value), Unsafe.As<T, int>(ref comparand)) == Unsafe.As<T, int>(ref comparand) :
+                typeof(T) == typeof(uint) ? Interlocked.CompareExchange(ref Unsafe.As<T, uint>(ref location), Unsafe.As<T, uint>(ref value), Unsafe.As<T, uint>(ref comparand)) == Unsafe.As<T, uint>(ref comparand) :
+                typeof(T) == typeof(long) ? Interlocked.CompareExchange(ref Unsafe.As<T, long>(ref location), Unsafe.As<T, long>(ref value), Unsafe.As<T, long>(ref comparand)) == Unsafe.As<T, long>(ref comparand) :
+                typeof(T) == typeof(ulong) ? Interlocked.CompareExchange(ref Unsafe.As<T, ulong>(ref location), Unsafe.As<T, ulong>(ref value), Unsafe.As<T, ulong>(ref comparand)) == Unsafe.As<T, ulong>(ref comparand) :
+                typeof(T) == typeof(nint) ? Interlocked.CompareExchange(ref Unsafe.As<T, nint>(ref location), Unsafe.As<T, nint>(ref value), Unsafe.As<T, nint>(ref comparand)) == Unsafe.As<T, nint>(ref comparand) :
+                typeof(T) == typeof(nuint) ? Interlocked.CompareExchange(ref Unsafe.As<T, nuint>(ref location), Unsafe.As<T, nuint>(ref value), Unsafe.As<T, nuint>(ref comparand)) == Unsafe.As<T, nuint>(ref comparand) :
+                throw new UnreachableException();
+
+            // The worker body. Each worker will execute this same body.
+            Func<object, Task> taskBody = static async o =>
+            {
+                var state = (ForEachState<T>)o;
+                bool launchedNext = false;
+
+#pragma warning disable CA2007 // Explicitly don't use ConfigureAwait, as we want to perform all work on the specified scheduler that's now current
+                try
+                {
+                    // Continue to loop while there are more elements to be processed.
+                    while (!state.Cancellation.IsCancellationRequested)
+                    {
+                        // Get the next element from the enumerator. For some types, we can get the next element with just
+                        // interlocked operations, avoiding the need to take a lock.  For other types, we need to take a lock.
+                        T element;
+                        if (Interlockable())
+                        {
+                            TryAgain:
+                            element = state.NextAvailable;
+                            if (element >= state.ToExclusive)
+                            {
+                                break;
+                            }
+
+                            if (!CompareExchange(ref state.NextAvailable, element + T.One, element))
+                            {
+                                goto TryAgain;
```

Comment:
Is this skipping to the next element if the next available element is not the same as the element + the next element?

I'm not sure I understand the purpose, or if I explained it correctly. Would you mind correcting me / explaining these lines?

Reply:
It's incrementing by 1 in a thread-safe manner, as multiple threads could all be racing to increment at the same time.